### PR TITLE
device: abstract API to walk dependencies

### PR DESCRIPTION
**Pull request by pabigot**
It's currently anticipated that we will need to support runtime registration of device dependencies.  In these cases there will not be a contiguous sequence that identifies all dependencies, as is the case with devicetree-defined dependencies.  So we'll need an API that is more flexible.

Add a visitor-pattern API to iterate over the devices, and switch to using this in existing code that wants to operate on dependency information.

### CODE REVIEW

**Comment by pabigot** [comment]
I haven't proposed an API to add runtime dependencies because it's not clear what we need to support.  Options include a range that covers at least:

* a single array of device pointers that will never change, is constructed in the device driver implementation, and registered with `device_state` once through API TBD and is never changed; to
* a linked list of devices that can be changed by the device or any other system component at any time.

The flexible alternatives introduce complexity in that walking the list requires locks, which may interfere with what the visitor is allowed to do.  We have not determined the context and privileges that power management requires of hooks, so best to leave this as theoretical until we do.

File: include/device.h [hunk 1]
```diff
@@ -427,6 +427,24 @@ device_from_handle(device_handle_t dev_handle)
 	return dev;
 }
 
+/**
+ * @brief Prototype for functions used when iterating over a set of devices.
+ *
+ * Such a function may be used in API that identifies a set of devices and
+ * provides an visitor API supporting caller-specific interaction with each
```
Comment (mbolivar-nordic) [inline include/device.h:434]:
```suggestion
 * provides a visitor API supporting caller-specific interaction with each
```

File: include/device.h [hunk 2]
```diff
@@ -427,6 +427,24 @@ device_from_handle(device_handle_t dev_handle)
...
+ * The visit is said to succeed if the visitor returns a non-negative value.
+ *
+ * @param dev a device in the set being iterated
+ *
+ * @param context state used to support the visitor function
+ *
+ * @return A non-negative number to allow walking to continue, and a negative
+ * error code to case the iteration to stop.
+ */
+typedef int (*device_visitor_cb)(const struct device *dev, void *context);
```
Comment (mbolivar-nordic) [inline include/device.h:446]: Looking through the other callback typedefs in the driver APIs, most of them seem to be named `foo_callback_t`.

Did you consider that and reject it? If so, why?

File: include/device.h [hunk 3]
```diff
@@ -460,6 +478,43 @@ device_required_handles_get(const struct device *dev,
 	return rv;
 }
 
+/**
+ * @brief Visit every device on with the provided device directly requires.
```
Comment (mbolivar-nordic) [inline include/device.h:482]:
```suggestion
 * @brief Visit every device that @p dev directly requires.
```

File: include/device.h [hunk 4]
```diff
@@ -460,6 +478,43 @@ device_required_handles_get(const struct device *dev,
...
+ *
+ * Zephyr maintains information about which devices are directly required by
+ * another device; for example an I2C-based sensor driver will require an I2C
+ * controller for communication.  Required devices can derive from
+ * statically-defined devicetree relationships or by dependencies registered
```
Comment (mbolivar-nordic) [inline include/device.h:487]:
```suggestion
 * statically-defined devicetree relationships or dependencies registered
```

File: tests/lib/devicetree/devices/src/main.c [hunk 5]
```diff
@@ -74,23 +74,51 @@ static bool check_handle(device_handle_t hdl,
 	return false;
 }
 
+struct requires_context {
+	uint8_t ndevs;
+	const struct device *rdevs[3];
```
Comment (mbolivar-nordic) [inline tests/lib/devicetree/devices/src/main.c:79]: Why 3, here and below; isn't 2 enough? I know 3 isn't an error, but I get weirded out by oversized buffers like this; makes me feel like I'm missing something. Am I?

**Review by mbolivar-nordic — COMMENTED** [review]
A few questions and docstring nits.

File: include/device.h (hunk 2 above)
Comment (pabigot) [reply include/device.h]: `device_visitor_callback_t` seemed a little long.

`git grep -P '_cb\b'| grep ':typedef'` shows 26 with `_cb`, 35 with `_callback_t`, though I agree device headers tend toward the latter

Changed.

File: tests/lib/devicetree/devices/src/main.c (hunk 5 above)
Comment (pabigot) [reply tests/lib/devicetree/devices/src/main.c]: I had an idea about checking that it didn't write more than was necessary, but didn't follow through.  On reconsideration that doesn't seem worth doing.

**Review by pabigot — COMMENTED** [review]
@mbolivar-nordic comments addressed

**Comment by mbolivar-nordic** [comment]
Why was this removed from the tests related to TEST_DEVA?

```c
zassert_true((ctx.rdevs[0] == device_from_handle(DEV_HDL(TEST_GPIO)))
		      || (ctx.rdevs[1] == device_from_handle(DEV_HDL(TEST_GPIO))),
		     NULL);
```

**Comment by mbolivar-nordic** [comment]
> (Oddly [#32547 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/32547#issuecomment-785300894) is the second comment I've encountered today where there's no place to reply in-thread. Maybe github changed something.)

I posted it as a top level comment.

**Review by mbolivar-nordic — APPROVED** [review]

**Review by ceolin — APPROVED** [review]

**Review by nordic-krch — APPROVED** [review]

File: include/device.h [hunk 6]
```diff
@@ -498,6 +516,43 @@ device_required_handles_get(const struct device *dev,
...
+ * dependency set.  This parameter must not be null.
+ *
+ * @param context state that is passed through to the visitor function.  This
+ * parameter may be null if @p visitor tolerates a null @p context.
+ *
+ * @return The number of devices that were visited if all visits succeed, or
+ * the negative value returned from the first visit that did not succeed.
+ */
+int device_visit_required(const struct device *dev,
+			  device_visitor_callback_t visitor,
```
Comment (nashif) [inline include/device.h:553]: visitor here is a callback, can we call this parameter `visitor_cb` instead to make it clear. This seems to be the de facto standard in the tree.

File: kernel/device.c [hunk 7]
```diff
@@ -162,6 +162,28 @@ bool z_device_ready(const struct device *dev)
 	return dev->state->initialized && (dev->state->init_res == 0);
 }
 
+int device_visit_required(const struct device *dev,
```
Comment (nashif) [inline kernel/device.c:165]: what made you go for `visit`? why not `iterate` or `walk` or `foreach`...? Visit just sounds weird.
Not an official statistic, but looking across the codebase, for this type of thing we use `_foreach`.

**Review by nashif — CHANGES_REQUESTED** [review]

File: kernel/device.c (hunk 7 above)
Comment (pabigot) [reply kernel/device.c:165]: Do you want:
* `device_foreach_required`
* `device required_foreach`
* `device_for_each_required`
* something else?

Comment (nashif) [reply kernel/device.c:165]: `device required_foreach`

**Review by nashif — APPROVED** [review]

**Pull request merged**